File: E-Commerce App Project (Tabbed)/Classes/AccountOperations.m
//
//  AccountOperations.m
//  E-Commerce App Project (Tabbed)
//
//  Created by Rony Banik on 13/11/18.
//  Copyright © 2018 Rony Banik. All rights reserved.
//

#import "AccountOperations.h"
#import <CommonCrypto/CommonCrypto.h>


@interface AccountOperations()

@property (strong, nonatomic) NSMutableArray* itemArray;

@end

@implementation AccountOperations

@synthesize isOfflineMode;

+ (AccountOperations *)sharedInstance
{
    static AccountOperations*  _sharedAccountOperations;
    
    static dispatch_once_t once;
    dispatch_once(&once,^{
        _sharedAccountOperations = [[AccountOperations alloc] init];
        _sharedAccountOperations.isOfflineMode = NO;
    });
    
    return _sharedAccountOperations;
}

- (NSDictionary *)getDummyUserData {
    return @{
        @"firstName": @"Demo",
        @"lastName": @"User",
        @"usersEmail": @"demo@example.com",
        @"phone": @"555-123-4567",
        @"address": @"123 Demo Street, Sample City",
        @"userExist": @"TRUE",
        @"RequestExecuted": @"TRUE",
        @"actionRequest": @"CHECK_USER_LOGIN"
    };
}

- (void)showOfflineNotification:(UIViewController *)viewController {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Offline Mode"
                                                                   message:@"Unable to connect to server. The app is running in offline mode with demo data."
                                                            preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK"
                                                       style:UIAlertActionStyleDefault
                                                     handler:nil];
    [alert addAction:okAction];
    [viewController presentViewController:alert animated:YES completion:nil];
}

- (void)handleOfflineModeWithAction:(NSString *)actionRequest callback:(void (^)(NSError *error, BOOL success, NSString* customErrorMessage))callback {
    self.isOfflineMode = YES;
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    
    if ([actionRequest isEqualToString:@"CHECK_USER_LOGIN"]) {
        NSDictionary *dummyData = [self getDummyUserData];
        [defaults setBool:YES forKey:@"SeesionUserLoggedIN"];
        [defaults setObject:dummyData[@"usersEmail"] forKey:@"SessionLoggedInuserEmail"];
        [defaults setObject:dummyData forKey:@"LoggedInUsersDetail"];
        [defaults synchronize];
        callback(nil, YES, @"Offline Mode: Logged in with demo account");
    } else if ([actionRequest isEqualToString:@"REGISTER_USER"]) {
        callback(nil, YES, @"Offline Mode: Registration simulated successfully");
    } else if ([actionRequest isEqualToString:@"EDIT_USER"]) {
        callback(nil, YES, @"Offline Mode: Profile update simulated successfully");
    } else {
        callback(nil, YES, @"Offline Mode: Operation completed with demo data");
    }
}

- (NSString *)sha1:(NSString *)str {
    const char *cStr = [str UTF8String];
    unsigned char result[CC_SHA1_DIGEST_LENGTH];
    CC_SHA1(cStr, (CC_LONG)strlen(cStr), result);
    NSString *s = [NSString  stringWithFormat:@"%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x",
                   result[0], result[1], result[2], result[3], result[4],
                   result[5], result[6], result[7],
                   result[8], result[9], result[10], result[11], result[12],
                   result[13], result[14], result[15],
                   result[16], result[17], result[18], result[19]
                   ];
    
    return s;
}
- ( NSURLSession * )getURLSession
{
    static NSURLSession *session = nil;
    static dispatch_once_t onceToken;
    dispatch_once( &onceToken,
                  ^{
                      NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
                      session = [NSURLSession sessionWithConfiguration:configuration];
                  } );
    return session;
}

-(void)sendRequestToServer:(NSDictionary *)dataToSend callback:(void (^)(NSError *error, BOOL success, NSString* customErrorMessage))callback{
    
    NSString *actionRequest = dataToSend[@"actionRequest"];
    
    NSError* jsonSerializationError;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dataToSend options:0 error:&jsonSerializationError];
    if (!jsonData) {
        //NSLog(@"Got jsonSerailError: %@",jsonSerializationError);
        [self handleOfflineModeWithAction:actionRequest callback:callback];
        return;
    }
    
    NSString *jsonString;
    jsonString = [[NSString alloc]initWithData:jsonData encoding:NSUTF8StringEncoding];
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
    NSData* requestData = [NSData dataWithBytes:[jsonString UTF8String] length:[jsonString lengthOfBytesUsingEncoding:NSUTF8StringEncoding]];
    
    [request setURL:[NSURL URLWithString:@"https://apiforios.appendtech.com/urltosendrequestwithdata.php?InitialSecureKey:ououhkju59703373367639792F423F4528482B4D6251655468576D5A7134743777217A25iuiu"]];
    [request setHTTPMethod:@"POST"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Accept"];
    [request setValue:@"application/json; charset=UTF-8" forHTTPHeaderField:@"Content-Type"];
    [request setValue:@"theHeaderString" forHTTPHeaderField:@"AmrLagto"];
    [request setValue:[NSString stringWithFormat:@"%lu", (unsigned long)[requestData length]] forHTTPHeaderField:@"Content-Length"];
    [request setHTTPBody:requestData];
    
    __weak AccountOperations *weakSelf = self;
    NSURLSessionDataTask *task = [[self getURLSession] dataTaskWithRequest:request completionHandler:^( NSData *data, NSURLResponse *response, NSError *error )
              {
                  dispatch_async( dispatch_get_main_queue(),
                        ^{
                            // parse returned data
                            NSString *customErrorMessage;
                            if (error) {
                                // Network error - fall back to offline mode
                                [weakSelf handleOfflineModeWithAction:actionRequest callback:callback];
                                return;
                            }
                            else {
                                    // for Quick Testing
//                                    NSString *result = [[NSString alloc] initWithData:data encoding:NSASCIIStringEncoding];
//                                    NSLog(@"result of strign %@",result);
                                    
                                    NSData* jsonData = [NSData dataWithData:data];
                                    NSArray* jsonArray = [NSJSONSerialization JSONObjectWithData:jsonData options:NSJSONReadingAllowFragments error:nil];
                                    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *) response;
                                    NSDictionary *responseDictionary = [[NSDictionary alloc]initWithDictionary:(NSDictionary*)jsonArray];
                                    
                                    NSUserDefaults *defaults= [NSUserDefaults standardUserDefaults];
                                    if (([httpResponse statusCode] == 200)&& ([[responseDictionary valueForKey:@"RequestExecuted"] isEqualToString:@"TRUE"])) {
                                        NSLog(@"Response Back korse");
                                        if ([[responseDictionary valueForKey:@"actionRequest"] isEqualToString:@"REGISTER_USER"]) {
                                            if ([[responseDictionary valueForKey:@"NullFieldsFound"] isEqualToString:@"TRUE"]) {
                                                //form empty
                                                customErrorMessage = @"Every Form Field is required for Successful purchase";
                                                callback(error, YES, customErrorMessage);
                                                
                                            }else if ([[responseDictionary valueForKey:@"userRegistrationPassMisMatch"] isEqualToString:@"YES"]){
                                                //confirm Password and pass mismatch
                                                customErrorMessage = @"Password and Confirm Password must be same!";
                                                callback(error, YES, customErrorMessage);
                                            }else if([[responseDictionary valueForKey:@"userAlreadyRegistered"] isEqualToString:@"YES"]){
                                                //email already registered
                                                customErrorMessage = @"This email is already registered";
                                                callback(error, YES, customErrorMessage);
                                            }else{
                                                // user registered
                                                customErrorMessage = @"Registraition Successful";
                                                callback(error, YES, customErrorMessage);
                                            }
                                        }else if ([[responseDictionary valueForKey:@"actionRequest"] isEqualToString:@"CHECK_USER_LOGIN"]){
                                            if ([[responseDictionary valueForKey:@"userExist"] isEqualToString:@"TRUE"]) {
                                                [defaults setBool:YES forKey:@"SeesionUserLoggedIN"];
                                                [defaults setObject:[responseDictionary valueForKey:@"usersEmail"] forKey:@"SessionLoggedInuserEmail"];
                                                [defaults setObject:responseDictionary forKey:@"LoggedInUsersDetail"];
                                                [defaults synchronize];
                                                customErrorMessage = @"Login Successful";
                                                callback(error, YES, customErrorMessage);
                                            }else if ([[responseDictionary valueForKey:@"passwordMismatch"] isEqualToString:@"YES"]) {
                                                [defaults setBool:NO forKey:@"SeesionUserLoggedIN"];
                                                [defaults setObject:[responseDictionary valueForKey:@""] forKey:@"SessionLoggedInuserEmail"];
                                                [defaults synchronize];
                                                customErrorMessage = @"Wrong Password!";
                                                callback(error, YES, customErrorMessage);
                                                
                                            }else{
                                                // login failed
                                                [defaults setBool:NO forKey:@"SeesionUserLoggedIN"];
                                                [defaults setObject:[responseDictionary valueForKey:@""] forKey:@"SessionLoggedInuserEmail"];
                                                [defaults synchronize];
                                                customErrorMessage = @"User Does Not Exist!";
                                                callback(error, YES, customErrorMessage);
                                            }
                                        }else if([[responseDictionary valueForKey:@"actionRequest"] isEqualToString:@"EDIT_USER"]){
                                            if ([[responseDictionary valueForKey:@"NullFieldsFound"] isEqualToString:@"TRUE"]) {
                                                //form empty
                                                customErrorMessage = @"Every Field is Required!";
                                                callback(error, YES, customErrorMessage);
                                            }else if ([[responseDictionary valueForKey:@"userUpdatePassMisMatch"] isEqualToString:@"YES"]){
                                                //confirm Password and pass mismatch
                                                customErrorMessage = @"Password and Confirm Password must be same!";
                                                callback(error, YES, customErrorMessage);
                                            }else{
                                                // user updated
                                                customErrorMessage = @"Update Successful";
                                                callback(error, YES, customErrorMessage);
                                            }
                                        }else{
                                            //Nothing Happened
                                            customErrorMessage = @"Check Your Request!";
                                            callback(error, YES, customErrorMessage);
                                        }
                                        
                                }else{
                                    //request not executed but reached the server - fall back to offline mode
                                    [weakSelf handleOfflineModeWithAction:actionRequest callback:callback];
                                    }
                                }
                            } );
                  }];
    [task resume];
}
- (BOOL)validateEmailAccount:(NSString*)checkString
{
    BOOL stricterFilter = NO;
    NSString *stricterFilterString = @"[A-Z0-9a-z\\._%+-]+@([A-Za-z0-9-]+\\.)+[A-Za-z]{2,4}";
    NSString *laxString = @".+@([A-Za-z0-9-]+\\.)+[A-Za-z]{2}[A-Za-z]*";
    NSString *emailRegex = stricterFilter ? stricterFilterString : laxString;
    NSPredicate *emailTest = [NSPredicate predicateWithFormat:@"SELF MATCHES %@", emailRegex];
    return [emailTest evaluateWithObject:checkString];
}
@end

File: E-Commerce App Project (Tabbed)/Classes/AccountOperations.swift
//
//  AccountOperations.swift
//  E-Commerce App Project (Tabbed)
//
//  Converted to Swift
//

import Foundation
import UIKit
import CommonCrypto

@objc class AccountOperations: NSObject {
    @objc static let sharedInstance = AccountOperations()
    
    @objc var isOfflineMode: Bool = false
    
    private var itemArray: [Any] = []
    private var urlSession: URLSession?
    
    private override init() {
        super.init()
        isOfflineMode = false
    }
    
    @objc func getDummyUserData() -> [String: String] {
        return [
            "firstName": "Demo",
            "lastName": "User",
            "usersEmail": "demo@example.com",
            "phone": "555-123-4567",
            "address": "123 Demo Street, Sample City",
            "userExist": "TRUE",
            "RequestExecuted": "TRUE",
            "actionRequest": "CHECK_USER_LOGIN"
        ]
    }
    
    @objc func showOfflineNotification(_ viewController: UIViewController) {
        let alert = UIAlertController(
            title: "Offline Mode",
            message: "Unable to connect to server. The app is running in offline mode with demo data.",
            preferredStyle: .alert
        )
        let okAction = UIAlertAction(title: "OK", style: .default, handler: nil)
        alert.addAction(okAction)
        viewController.present(alert, animated: true, completion: nil)
    }
    
    private func handleOfflineMode(withAction actionRequest: String?, callback: @escaping (Error?, Bool, String?) -> Void) {
        isOfflineMode = true
        let defaults = UserDefaults.standard
        
        if actionRequest == "CHECK_USER_LOGIN" {
            let dummyData = getDummyUserData()
            defaults.set(true, forKey: "SeesionUserLoggedIN")
            defaults.set(dummyData["usersEmail"], forKey: "SessionLoggedInuserEmail")
            defaults.set(dummyData, forKey: "LoggedInUsersDetail")
            defaults.synchronize()
            callback(nil, true, "Offline Mode: Logged in with demo account")
        } else if actionRequest == "REGISTER_USER" {
            callback(nil, true, "Offline Mode: Registration simulated successfully")
        } else if actionRequest == "EDIT_USER" {
            callback(nil, true, "Offline Mode: Profile update simulated successfully")
        } else {
            callback(nil, true, "Offline Mode: Operation completed with demo data")
        }
    }
    
    @objc func sha1(_ str: String) -> String {
        let data = Data(str.utf8)
        var digest = [UInt8](repeating: 0, count: Int(CC_SHA1_DIGEST_LENGTH))
        data.withUnsafeBytes {
            _ = CC_SHA1($0.baseAddress, CC_LONG(data.count), &digest)
        }
        return digest.map { String(format: "%02x", $0) }.joined()
    }
    
    @objc func getURLSession() -> URLSession {
        if urlSession == nil {
            let configuration = URLSessionConfiguration.default
            urlSession = URLSession(configuration: configuration)
        }
        return urlSession!
    }
    
    @objc func sendRequest(toServer dataToSend: [String: Any], callback: @escaping (Error?, Bool, String?) -> Void) {
        let actionRequest = dataToSend["actionRequest"] as? String
        
        guard let jsonData = try? JSONSerialization.data(withJSONObject: dataToSend, options: []) else {
            handleOfflineMode(withAction: actionRequest, callback: callback)
            return
        }
        
        guard let jsonString = String(data: jsonData, encoding: .utf8),
              let url = URL(string: "https://apiforios.appendtech.com/urltosendrequestwithdata.php?InitialSecureKey:ououhkju59703373367639792F423F4528482B4D6251655468576D5A7134743777217A25iuiu") else {
            handleOfflineMode(withAction: actionRequest, callback: callback)
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/json; charset=UTF-8", forHTTPHeaderField: "Content-Type")
        request.setValue("theHeaderString", forHTTPHeaderField: "AmrLagto")
        
        let requestData = jsonString.data(using: .utf8)!
        request.setValue("\(requestData.count)", forHTTPHeaderField: "Content-Length")
        request.httpBody = requestData
        
        let task = getURLSession().dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                guard let self = self else { return }
                
                if error != nil {
                    self.handleOfflineMode(withAction: actionRequest, callback: callback)
                    return
                }
                
                guard let data = data,
                      let httpResponse = response as? HTTPURLResponse,
                      let responseDictionary = try? JSONSerialization.jsonObject(with: data, options: .fragmentsAllowed) as? [String: Any] else {
                    self.handleOfflineMode(withAction: actionRequest, callback: callback)
                    return
                }
                
                let defaults = UserDefaults.standard
                
                if httpResponse.statusCode == 200 && (responseDictionary["RequestExecuted"] as? String) == "TRUE" {
                    let responseAction = responseDictionary["actionRequest"] as? String
                    
                    if responseAction == "REGISTER_USER" {
                        if (responseDictionary["NullFieldsFound"] as? String) == "TRUE" {
                            callback(nil, true, "Every Form Field is required for Successful purchase")
                        } else if (responseDictionary["userRegistrationPassMisMatch"] as? String) == "YES" {
                            callback(nil, true, "Password and Confirm Password must be same!")
                        } else if (responseDictionary["userAlreadyRegistered"] as? String) == "YES" {
                            callback(nil, true, "This email is already registered")
                        } else {
                            callback(nil, true, "Registraition Successful")
                        }
                    } else if responseAction == "CHECK_USER_LOGIN" {
                        if (responseDictionary["userExist"] as? String) == "TRUE" {
                            defaults.set(true, forKey: "SeesionUserLoggedIN")
                            defaults.set(responseDictionary["usersEmail"], forKey: "SessionLoggedInuserEmail")
                            defaults.set(responseDictionary, forKey: "LoggedInUsersDetail")
                            defaults.synchronize()
                            callback(nil, true, "Login Successful")
                        } else if (responseDictionary["passwordMismatch"] as? String) == "YES" {
                            defaults.set(false, forKey: "SeesionUserLoggedIN")
                            defaults.set("", forKey: "SessionLoggedInuserEmail")
                            defaults.synchronize()
                            callback(nil, true, "Wrong Password!")
                        } else {
                            defaults.set(false, forKey: "SeesionUserLoggedIN")
                            defaults.set("", forKey: "SessionLoggedInuserEmail")
                            defaults.synchronize()
                            callback(nil, true, "User Does Not Exist!")
                        }
                    } else if responseAction == "EDIT_USER" {
                        if (responseDictionary["NullFieldsFound"] as? String) == "TRUE" {
                            callback(nil, true, "Every Field is Required!")
                        } else if (responseDictionary["userUpdatePassMisMatch"] as? String) == "YES" {
                            callback(nil, true, "Password and Confirm Password must be same!")
                        } else {
                            callback(nil, true, "Update Successful")
                        }
                    } else {
                        callback(nil, true, "Check Your Request!")
                    }
                } else {
                    self.handleOfflineMode(withAction: actionRequest, callback: callback)
                }
            }
        }
        task.resume()
    }
    
    @objc func validateEmailAccount(_ checkString: String) -> Bool {
        let stricterFilter = false
        let stricterFilterString = "[A-Z0-9a-z\\._%+-]+@([A-Za-z0-9-]+\\.)+[A-Za-z]{2,4}"
        let laxString = ".+@([A-Za-z0-9-]+\\.)+[A-Za-z]{2}[A-Za-z]*"
        let emailRegex = stricterFilter ? stricterFilterString : laxString
        let emailTest = NSPredicate(format: "SELF MATCHES %@", emailRegex)
        return emailTest.evaluate(with: checkString)
    }
}
